dataplane/dplanerc/interface.go

```diff
@@ -582,118 +615,188 @@ func (ni *Reconciler) reconcile(ctx context.Context, config *oc.Interface) {
+
+	vlanIntf := &netlink.Vlan{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:        intfRefToDevName(intfRef),
+			ParentIndex: rootPort.hostifIfIndex,
+		},
+		VlanId:       int(config.GetSubinterface(intfRef.subintf).GetVlan().GetMatch().GetSingleTagged().GetVlanId()),
+		VlanProtocol: netlink.VLAN_PROTOCOL_8021Q,
+	}
+	if err := ni.ifaceMgr.LinkAdd(vlanIntf); err != nil {
+		log.Warningf("failed to add vlan intf: %v", err)
+	}
+
+	rifResp, err := ni.ifaceClient.CreateRouterInterface(ctx, &saipb.CreateRouterInterfaceRequest{
+		Switch:          ni.switchID,
+		Type:            saipb.RouterInterfaceType_ROUTER_INTERFACE_TYPE_SUB_PORT.Enum(),
+		PortId:          &rootPort.portID,
+		VirtualRouterId: proto.Uint64(0),
+	})
+	if err != nil {
+		log.Warningf("failed to add vlan intf: %v", err)
 	}
 
+	ni.ocInterfaceData[intfRef] = &interfaceData{
+		rifID:         rifResp.GetOid(),
+		hostifIfIndex: vlanIntf.Index,
+		hostifDevName: intfRefToDevName(intfRef),
+	}
+	sb := &ygnmi.SetBatch{}
+	gnmiclient.BatchUpdate(sb, ocpath.Root().Interface(intfRef.name).Subinterface(intfRef.subintf).Vlan().Match().SingleTagged().VlanId().State(), uint16(vlanIntf.VlanId))
+	if _, err := sb.Set(ctx, ni.c); err != nil {
+		log.Warningf("failed to set link status: %v", err)
+	}
+}
+
+func (ni *Reconciler) reconcileIPs(config, state *oc.Interface) {
 	type prefixPair struct {
 		cfgIP, stateIP *string
 		cfgPL, statePL *uint8
 	}
 
-	// Get all state IPs and their corresponding config IPs (if they exist).
-	var interfacePairs []*prefixPair
-	for _, addr := range state.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
-		pair := &prefixPair{
-			stateIP: addr.Ip,
-			statePL: addr.PrefixLength,
-		}
-		if pairAddr := config.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp()); pairAddr != nil {
-			pair.cfgIP = pairAddr.Ip
-			pair.cfgPL = pairAddr.PrefixLength
+	for idx := range config.Subinterface {
+		intfRef := ocInterface{
+			name:    config.GetName(),
+			subintf: idx,
 		}
-		interfacePairs = append(interfacePairs, pair)
-	}
-	for _, addr := range state.GetOrCreateSubinterface(0).GetOrCreateIpv6().Address {
-		pair := &prefixPair{
-			stateIP: addr.Ip,
-			statePL: addr.PrefixLength,
-		}
-		if pairAddr := config.GetSubinterface(0).GetIpv6().GetAddress(addr.GetIp()); pairAddr != nil {
-			pair.cfgIP = pairAddr.Ip
-			pair.cfgPL = pairAddr.PrefixLength
+		data, ok := ni.ocInterfaceData[intfRef]
+		if !ok {
+			log.Infof("skipping ip reconcilation for %+v", intfRef)
+			continue
 		}
-		interfacePairs = append(interfacePairs, pair)
-	}
 
-	// Get all config IPs and their corresponding state IPs (if they exist).
-	for _, addr := range config.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
-		pair := &prefixPair{
-			cfgIP: addr.Ip,
-			cfgPL: addr.PrefixLength,
+		// Get all state IPs and their corresponding config IPs (if they exist).
+		var interfacePairs []*prefixPair
+		for _, addr := range state.GetOrCreateSubinterface(idx).GetOrCreateIpv4().Address {
```

Comment:
Should this just be `Get` vs. `GetOrCreate` to avoid mutating the state data?

Reply:
this should be fixed once i migrate this to the new reconciler api